
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>proteometer.abundance &#8212; Proteometer  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=4ae1632d" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/custom.css" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=8a448e45"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=ccdb6887"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'autoapi/proteometer/abundance/index';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="proteometer.alignment" href="../alignment/index.html" />
    <link rel="prev" title="proteometer" href="../index.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Proteometer  documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../index.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/PhenoMeters/proteometer" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../../index.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/PhenoMeters/proteometer" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../index.html">proteometer</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">proteometer.abundance</a></li>
<li class="toctree-l2"><a class="reference internal" href="../alignment/index.html">proteometer.alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="../barcode/index.html">proteometer.barcode</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fasta/index.html">proteometer.fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lip/index.html">proteometer.lip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../lip_analysis/index.html">proteometer.lip_analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../normalization/index.html">proteometer.normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../params/index.html">proteometer.params</a></li>
<li class="toctree-l2"><a class="reference internal" href="../parse_metadata/index.html">proteometer.parse_metadata</a></li>
<li class="toctree-l2"><a class="reference internal" href="../peptide/index.html">proteometer.peptide</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ptm/index.html">proteometer.ptm</a></li>
<li class="toctree-l2"><a class="reference internal" href="../ptm_analysis/index.html">proteometer.ptm_analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quality_control_plots/index.html">proteometer.quality_control_plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="../residue/index.html">proteometer.residue</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rollup/index.html">proteometer.rollup</a></li>
<li class="toctree-l2"><a class="reference internal" href="../stats/index.html">proteometer.stats</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils/index.html">proteometer.utils</a></li>
</ul>
</details></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../../../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../../index.html" class="nav-link">API Reference</a></li>
    
    
    <li class="breadcrumb-item"><a href="../index.html" class="nav-link">proteometer</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">proteometer.abundance</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-proteometer.abundance">
<span id="proteometer-abundance"></span><h1>proteometer.abundance<a class="headerlink" href="#module-proteometer.abundance" title="Link to this heading">#</a></h1>
<section id="functions">
<h2>Functions<a class="headerlink" href="#functions" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#proteometer.abundance.get_prot_abund_scalars" title="proteometer.abundance.get_prot_abund_scalars"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_prot_abund_scalars</span></code></a>(→ dict[str, float])</p></td>
<td><p>Return a dictionary of protein abundance scalars for the given pairwise t-test.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#proteometer.abundance.prot_abund_correction" title="proteometer.abundance.prot_abund_correction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prot_abund_correction</span></code></a>(→ pandas.DataFrame)</p></td>
<td><p>Perform protein abundance correction based on the provided parameters.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#proteometer.abundance.prot_abund_correction_sig_only" title="proteometer.abundance.prot_abund_correction_sig_only"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prot_abund_correction_sig_only</span></code></a>(→ pandas.DataFrame)</p></td>
<td><p>Adjusts peptide abundance values based on protein abundance scalars for</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#proteometer.abundance.prot_abund_correction_matched" title="proteometer.abundance.prot_abund_correction_matched"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prot_abund_correction_matched</span></code></a>(→ pandas.DataFrame)</p></td>
<td><p>Correct the peptide abundance data using the protein abundance values.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#proteometer.abundance.global_prot_normalization_and_stats" title="proteometer.abundance.global_prot_normalization_and_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">global_prot_normalization_and_stats</span></code></a>(→ pandas.DataFrame)</p></td>
<td><p>Perform global protein normalization and statistical analysis.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="proteometer.abundance.get_prot_abund_scalars">
<span class="sig-prename descclassname"><span class="pre">proteometer.abundance.</span></span><span class="sig-name descname"><span class="pre">get_prot_abund_scalars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairwise_ttest_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pval'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)"><span class="pre">dict</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><span class="pre">float</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference external" href="https://github.com/PhenoMeters/proteometer/blob/main/proteometer/abundance.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#proteometer.abundance.get_prot_abund_scalars" title="Link to this definition">#</a></dt>
<dd><p>Return a dictionary of protein abundance scalars for the given pairwise t-test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot</strong> (<em>pd.DataFrame</em>) – DataFrame containing protein-level data.</p></li>
<li><p><strong>pairwise_ttest_name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Name of the pairwise t-test.</p></li>
<li><p><strong>sig_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – Type of significance metric to use for filtering.
Defaults to “pval”.</p></li>
<li><p><strong>sig_thr</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><em>optional</em>) – Threshold for significance filtering. Defaults to 0.05.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of protein abundance scalars.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.13)">dict</a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)">float</a>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="proteometer.abundance.prot_abund_correction">
<span class="sig-prename descclassname"><span class="pre">proteometer.abundance.</span></span><span class="sig-name descname"><span class="pre">prot_abund_correction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pept</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">par</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="../params/index.html#proteometer.params.Params" title="proteometer.params.Params"><span class="pre">proteometer.params.Params</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns_to_correct</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairwise_ttest_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><span class="pre">proteometer.stats.TTestGroup</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_tt_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="reference external" href="https://github.com/PhenoMeters/proteometer/blob/main/proteometer/abundance.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#proteometer.abundance.prot_abund_correction" title="Link to this definition">#</a></dt>
<dd><p>Perform protein abundance correction based on the provided parameters.</p>
<p>This function applies either paired or unpaired sample abundance correction
depending on the <code class="code docutils literal notranslate"><span class="pre">abundance_correction_paired_samples</span></code> attribute of the <code class="code docutils literal notranslate"><span class="pre">par</span></code> parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pept</strong> (<em>pd.DataFrame</em>) – A DataFrame containing peptide-level data.</p></li>
<li><p><strong>prot</strong> (<em>pd.DataFrame</em>) – A DataFrame containing protein-level data.</p></li>
<li><p><strong>par</strong> (<a class="reference internal" href="../params/index.html#proteometer.params.Params" title="proteometer.params.Params"><em>Params</em></a>) – A parameter object containing configuration for abundance correction.</p></li>
<li><p><strong>columns_to_correct</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Columns to correct for paired sample abundance correction.
Required if <code class="code docutils literal notranslate"><span class="pre">par.abundance_correction_paired_samples</span></code> is True.</p></li>
<li><p><strong>pairwise_ttest_groups</strong> (<em>Iterable</em><em>[</em><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><em>stats.TTestGroup</em></a><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Groups for pairwise t-tests in unpaired sample abundance correction.
Required if <code class="code docutils literal notranslate"><span class="pre">par.abundance_correction_paired_samples</span></code> is False.</p></li>
<li><p><strong>non_tt_cols</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Columns that should not be included in the t-test correction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame with corrected protein abundances.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.13)"><strong>ValueError</strong></a> – If <code class="code docutils literal notranslate"><span class="pre">columns_to_correct</span></code> is not provided for paired sample correction.</p></li>
<li><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.13)"><strong>ValueError</strong></a> – If <code class="code docutils literal notranslate"><span class="pre">pairwise_ttest_groups</span></code> is not provided for unpaired sample correction.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="proteometer.abundance.prot_abund_correction_sig_only">
<span class="sig-prename descclassname"><span class="pre">proteometer.abundance.</span></span><span class="sig-name descname"><span class="pre">prot_abund_correction_sig_only</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pept</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairwise_ttest_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><span class="pre">proteometer.stats.TTestGroup</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uniprot_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'pval'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="reference external" href="https://github.com/PhenoMeters/proteometer/blob/main/proteometer/abundance.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#proteometer.abundance.prot_abund_correction_sig_only" title="Link to this definition">#</a></dt>
<dd><p>Adjusts peptide abundance values based on protein abundance scalars for
significant pairwise t-test groups.</p>
<p>This function iterates over a collection of pairwise t-test groups, computes
or retrieves protein abundance scalars, and applies these scalars to adjust
the peptide abundance values for the specified treatment samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pept</strong> (<em>pd.DataFrame</em>) – DataFrame containing peptide-level data. Must include
a column corresponding to <code class="code docutils literal notranslate"><span class="pre">uniprot_col</span></code> for mapping protein identifiers.</p></li>
<li><p><strong>prot</strong> (<em>pd.DataFrame</em>) – DataFrame containing protein-level data. Must include
columns for protein abundance scalars or data required to compute them.</p></li>
<li><p><strong>pairwise_ttest_groups</strong> (<em>Iterable</em><em>[</em><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><em>stats.TTestGroup</em></a><em>]</em>) – An iterable of
TTestGroup objects, each representing a pairwise t-test group with
associated metadata (e.g., labels and treatment samples).</p></li>
<li><p><strong>uniprot_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Column name in <code class="code docutils literal notranslate"><span class="pre">pept</span></code> that contains UniProt identifiers
for mapping to protein abundance data.</p></li>
<li><p><strong>sig_type</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>, </em><em>optional</em>) – Type of significance metric to use for filtering
(e.g., “pval” for p-value or “adj-p” for adjusted p-value). Defaults to “pval”.</p></li>
<li><p><strong>sig_thr</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.13)"><em>float</em></a><em>, </em><em>optional</em>) – Threshold for significance filtering. Only proteins
meeting this threshold will have their abundance scalars applied. Defaults to 0.05.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Updated <code class="code docutils literal notranslate"><span class="pre">pept</span></code> DataFrame with adjusted abundance values for</dt><dd><p>treatment samples and additional columns for protein abundance scalars.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="proteometer.abundance.prot_abund_correction_matched">
<span class="sig-prename descclassname"><span class="pre">proteometer.abundance.</span></span><span class="sig-name descname"><span class="pre">prot_abund_correction_matched</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pept</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">columns_to_correct</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uniprot_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_tt_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.13)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="reference external" href="https://github.com/PhenoMeters/proteometer/blob/main/proteometer/abundance.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#proteometer.abundance.prot_abund_correction_matched" title="Link to this definition">#</a></dt>
<dd><p>Correct the peptide abundance data using the protein abundance values.</p>
<p>This function takes the peptide data and corrects the intensity values
for each peptide using the protein abundance values from the protein
data. The correction is only applied to the treatment samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pept</strong> (<em>pd.DataFrame</em>) – A DataFrame containing peptide-level data.</p></li>
<li><p><strong>prot</strong> (<em>pd.DataFrame</em>) – A DataFrame containing protein-level data.</p></li>
<li><p><strong>columns_to_correct</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – Columns to correct for protein
abundance changes. Must be shared by <code class="code docutils literal notranslate"><span class="pre">pept</span></code> and <code class="code docutils literal notranslate"><span class="pre">prot</span></code>.</p></li>
<li><p><strong>uniprot_col</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a>) – Column name for the Uniprot ID in both <code class="code docutils literal notranslate"><span class="pre">pept</span></code> and <code class="code docutils literal notranslate"><span class="pre">prot</span></code>.</p></li>
<li><p><strong>non_tt_cols</strong> (<em>Iterable</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Columns that should not
be included in the abundance correction. Must be shared by <code class="code docutils literal notranslate"><span class="pre">pept</span></code> and <code class="code docutils literal notranslate"><span class="pre">prot</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Updated <code class="code docutils literal notranslate"><span class="pre">pept</span></code> DataFrame with adjusted abundance values</dt><dd><p>for treatment samples and additional columns for protein abundance
scalars.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="proteometer.abundance.global_prot_normalization_and_stats">
<span class="sig-prename descclassname"><span class="pre">proteometer.abundance.</span></span><span class="sig-name descname"><span class="pre">global_prot_normalization_and_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">global_prot</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">int_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anova_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><span class="pre">list</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairwise_ttest_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable" title="(in Python v3.13)"><span class="pre">collections.abc.Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><span class="pre">proteometer.stats.TTestGroup</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metadata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">pandas.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">par</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="../params/index.html#proteometer.params.Params" title="proteometer.params.Params"><span class="pre">proteometer.params.Params</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">pandas.DataFrame</span></span></span><a class="reference external" href="https://github.com/PhenoMeters/proteometer/blob/main/proteometer/abundance.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#proteometer.abundance.global_prot_normalization_and_stats" title="Link to this definition">#</a></dt>
<dd><p>Perform global protein normalization and statistical analysis.</p>
<p>This function applies normalization and statistical tests to global proteomics
data. It handles both median normalization and batch correction, depending on
the parameters provided in the <code class="code docutils literal notranslate"><span class="pre">par</span></code> object. It also performs ANOVA and pairwise t-tests.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>global_prot</strong> (<em>pd.DataFrame</em>) – DataFrame containing global protein-level data.</p></li>
<li><p><strong>int_cols</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – List of column names representing intensity data to normalize.</p></li>
<li><p><strong>anova_cols</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.13)"><em>list</em></a><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.13)"><em>str</em></a><em>]</em>) – List of column names for ANOVA analysis.</p></li>
<li><p><strong>pairwise_ttest_groups</strong> (<em>Iterable</em><em>[</em><a class="reference internal" href="../stats/index.html#proteometer.stats.TTestGroup" title="proteometer.stats.TTestGroup"><em>stats.TTestGroup</em></a><em>]</em>) – Iterable of TTestGroup objects
for performing pairwise t-tests (each defines a control-treatment pair).</p></li>
<li><p><strong>metadata</strong> (<em>pd.DataFrame</em>) – DataFrame containing metadata for batch correction and
ANOVA analysis.</p></li>
<li><p><strong>par</strong> (<a class="reference internal" href="../params/index.html#proteometer.params.Params" title="proteometer.params.Params"><em>Params</em></a>) – Parameter object containing configuration for normalization and
statistical analysis.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The normalized and statistically analyzed global protein data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../index.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">proteometer</p>
      </div>
    </a>
    <a class="right-next"
       href="../alignment/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">proteometer.alignment</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functions">Functions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-contents">Module Contents</a><ul class="visible nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#proteometer.abundance.get_prot_abund_scalars"><code class="docutils literal notranslate"><span class="pre">get_prot_abund_scalars()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#proteometer.abundance.prot_abund_correction"><code class="docutils literal notranslate"><span class="pre">prot_abund_correction()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#proteometer.abundance.prot_abund_correction_sig_only"><code class="docutils literal notranslate"><span class="pre">prot_abund_correction_sig_only()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#proteometer.abundance.prot_abund_correction_matched"><code class="docutils literal notranslate"><span class="pre">prot_abund_correction_matched()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#proteometer.abundance.global_prot_normalization_and_stats"><code class="docutils literal notranslate"><span class="pre">global_prot_normalization_and_stats()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../../../_sources/autoapi/proteometer/abundance/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, PNNL.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>